cluster/router/script/instance/js_instance.go

```diff
@@ -0,0 +1,203 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package instance
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"sync/atomic"
+)
+
+import (
+	"dubbo.apache.org/dubbo-go/v3/protocol"
+
+	"github.com/dop251/goja"
+)
+
+const (
+	jsScriptResultName = `__go_program_result`
+	jsScriptPrefix     = "\n" + jsScriptResultName + ` = `
+)
+
+type jsInstances struct {
+	insPool *sync.Pool // store *goja.runtime
+	pgLock  sync.RWMutex
+	program map[string]*program // rawScript to compiledProgram
+}
+
+type jsInstance struct {
+	rt *goja.Runtime
+}
+
+type program struct {
+	pg    *goja.Program
+	count int32
+}
+
+func newProgram(pg *goja.Program) *program {
+	return &program{
+		pg:    pg,
+		count: 1,
+	}
+}
+
+func (p *program) addCount(i int) int {
+	return int(atomic.AddInt32(&p.count, int32(i)))
+}
+
+func newJsInstances() *jsInstances {
+	return &jsInstances{
+		insPool: &sync.Pool{New: func() any {
+			return newJsInstance()
+		}},
+	}
+}
+
+func (i *jsInstances) Run(rawScript string, invokers []protocol.Invoker, invocation protocol.Invocation) ([]protocol.Invoker, error) {
+	i.pgLock.RLock()
+	pg, ok := i.program[rawScript]
+	i.pgLock.RUnlock()
+
+	if !ok || len(invokers) == 0 {
+		return invokers, nil
+	}
+	matcher := i.insPool.Get().(*jsInstance)
+
+	packInvokers := make([]protocol.Invoker, 0, len(invokers))
+	for _, invoker := range invokers {
+		packInvokers = append(packInvokers, newScriptInvokerImpl(invoker))
+	}
+
+	ctx := invocation.GetAttachmentAsContext()
+	matcher.initCallArgs(packInvokers, invocation, ctx)
+	matcher.initReplyVar()
+	scriptRes, err := matcher.runScript(pg.pg)
+	if err != nil {
+		return invokers, err
+	}
+	invocation.MergeAttachmentFromContext(ctx)
+
+	rtInvokersArr, ok := scriptRes.(*goja.Object).Export().([]interface{})
+	if !ok {
+		return invokers, fmt.Errorf("script result is not array , script return type: %s", reflect.ValueOf(scriptRes.(*goja.Object).Export()).String())
+	}
+
+	result := make([]protocol.Invoker, 0, len(rtInvokersArr))
+	for _, res := range rtInvokersArr {
+		if i, ok := res.(*scriptInvokerWrapper); ok {
+			i.setRanMode()
+			result = append(result, res.(protocol.Invoker))
+		} else {
+			return invokers, fmt.Errorf("invalid element type , it should be (*scriptInvokerWrapper) , but type is : %s)", reflect.TypeOf(res).String())
+		}
+	}
+	return result, nil
+}
+
+func (i *jsInstances) Compile(key, rawScript string) error {
+	var (
+		ok bool
+		pg *program
+	)
+
+	i.pgLock.RLock()
+	pg, ok = i.program[rawScript]
+	i.pgLock.RUnlock()
+
+	if ok {
+		pg.addCount(1)
+		return nil
+	} else {
+
+		newPg, err := goja.Compile(key+`_jsScriptRoute`, jsScriptPrefix+rawScript, true)
+		if err != nil {
+			return err
+		}
+
+		i.pgLock.Lock()
+		// double check to avoid race
+		if pg, ok = i.program[rawScript]; ok {
+			pg.addCount(1)
+		} else {
+			i.program[rawScript] = newProgram(newPg)
+		}
+		i.pgLock.Unlock()
+		return nil
+	}
```

Comment:
The lock used here is a bit strange, try the following steps:
1. check without lock
2. if exists, return
3. if not exists, lock
   3.1 check again
   3.2 if exists return
   3.3 if not exists compile

Reply:
i will rewite it.

Reply:
I see. I thought the following lock was not necessary, it seems to be necessary in this case.

```
i.pgLock.RLock()
pg, ok = i.program[rawScript]
i.pgLock.RUnlock()
```